Pass compile mode to the custom build script

nyurik opened the pull request:
Closes #4001

This PR adds code to track and pass compile mode to the build scripts. Since the unit's mode is always `RunCustomBuild`, I had to add tracking for the parent unit's compile mode, and pass it around.

### TODO:
  - [ ] env var naming. I used `CARGO_MODE`, but perhaps another name is better? (e.g. `CARGO_CMD`, `CARGO_COMMAND`, ... ?)
  - [ ] env var values naming. I used: `test`, `build`, `check`, `bench`, `doc`, `doctest`, and `docscrape`. The last two might need to be substituted with `doc`, or panic on them?
  - [ ] Should the bool values in `Check` and `Doc` modes be shown? Both values are currently ignored.
  - [ ] figure out why `cargo bench` sets env var to `test`
  - [ ] figure out why unit tests are failing - doesn't seem to be related to changes.
  - [ ] add unit tests

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Eh2406 (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

jyn514 commented:
> Test, Build, Check_test/Check, Bench, Doc_with_deps/Doc, Doctest, Docscrape, RunCustomBuild

nit: I think it would be better to use snake_case for these, the same way that they're named on the command line.

I also think there doesn't need to be so many modes; just `check`, `build`, `test`, and `doc` seem enough. `doc_scrape` in particular shouldn't be passed because it's unstable and build scripts shouldn't depend on it.

jyn514 commented on src/cargo/core/compiler/custom_build.rs:
```diff
@@ -201,6 +202,31 @@ fn build_work(cx: &mut Context<'_, '_>, unit: &Unit) -> CargoResult<Job> {
+                }
+                CompileMode::Bench => "Bench",
+                CompileMode::Doc { deps } => {
+                    if deps {
```
This shouldn't matter to the build script, no dependencies will be built *after* it runs.

jyn514 commented on src/cargo/core/compiler/unit.rs:
```diff
@@ -52,6 +52,8 @@ pub struct UnitInner {
     /// The "mode" this unit is being compiled for. See [`CompileMode`] for more details.
     pub mode: CompileMode,
+    /// The "mode" of the root unit. Required when unit's mode is `CompileMode::RunCustomBuild`
+    pub root_mode: CompileMode,
```
Rather than adding an additional field you don't always need, I think it might make sense to add a field to the RunCustomBuild unit:
```rust
enum CompileMode {
    RunCustomBuild { root_mode: Box<CompileMode> },
}
```

nyurik replied:
@jyn514 that's an awesome idea, thanks!  One issue -- the enum implements `Copy` trait which cannot be auto-implemented. I wonder if i can get away with the `Arc` instead?

jyn514 reviewed (commented):
I left a few suggestions, but I defer to Ehuss's expertise 😁 I did see that this doesn't have any dependency tracking - I think there needs to be some logic to rerun the build script when the root_mode changes.

nyurik commented:
> I left a few suggestions, but I defer to Ehuss's expertise grin I did see that this doesn't have any dependency tracking - I think there needs to be some logic to rerun the build script when the root_mode changes.

@jyn514 thx again for the feedback, I switched to `&'static str` instead of the Box - this way `Copy` trait continues to work. Also renamed the strings. There is still a question of what to do with `Doctest` and `Docscrape` (see code)

bors commented:
:umbrella: The latest upstream changes (presumably #9992) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #10434) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ehuss commented:
Closing as this PR has become a bit stale.  Also, this appears to be a pretty major change that will be quite difficult to make as it will have an impact on every project using build scripts.  Unfortunately the Cargo team doesn't have the capacity to accept new features of this nature at this time.
